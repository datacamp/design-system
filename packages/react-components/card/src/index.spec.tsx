import '@testing-library/jest-dom/extend-expect';

import axeRender from '@datacamp/waffles-axe-render';
import { shadow } from '@datacamp/waffles-tokens/lib/future-tokens.json';
import * as React from 'react';

import Card, { elevationMap } from '.';

const testText = 'example content';

describe('<Card />', () => {
  describe('element type', () => {
    it('is specified using as prop', async () => {
      const { container } = await axeRender(
        <Card as="section">
          <p>{testText}</p>
        </Card>
      );

      expect((container.firstChild as HTMLElement).tagName).toEqual('SECTION');
    });

    it('is a div by default', async () => {
      const { container } = await axeRender(
        <Card>
          <p>{testText}</p>
        </Card>
      );

      expect((container.firstChild as HTMLElement).tagName).toEqual('DIV');
    });
  });

  describe('children', () => {
    it('is rendered inside the card', async () => {
      const { container, getByText } = await axeRender(
        <Card as="section">
          <p>{testText}</p>
        </Card>
      );
      const childElement = getByText(testText) as HTMLElement;

      expect(container.firstChild).toContainElement(childElement);
    });
  });

  describe('className', () => {
    it('is set on the card element', async () => {
      const testClass = 'test-className';
      const { container } = await axeRender(
        <Card className={testClass}>
          <p>{testText}</p>
        </Card>
      );

      expect(container.firstChild).toHaveClass(testClass);
    });
  });

  describe('elevation', () => {
    it('has the border box shadow by default', async () => {
      const { container } = await axeRender(
        <Card>
          <p>{testText}</p>
        </Card>
      );

      expect(container.firstChild).toHaveStyle(
        `box-shadow: ${shadow.border.value}`
      );
    });

    it('has the border box shadow for elevation = 0', async () => {
      const { container } = await axeRender(
        <Card elevation={0}>
          <p>{testText}</p>
        </Card>
      );

      expect(container.firstChild).toHaveStyle(
        `box-shadow: ${shadow.border.value}`
      );
    });

    it('has the small box shadow for elevation = 1', async () => {
      const { container } = await axeRender(
        <Card elevation={1}>
          <p>{testText}</p>
        </Card>
      );

      expect(container.firstChild).toHaveStyle(
        `box-shadow: 0px 0px 1px 0px rgba(61,66,81,0.3),${shadow.sm.value}`
      );
    });

    it('has the medium box shadow for elevation = 2', async () => {
      const { container } = await axeRender(
        <Card elevation={2}>
          <p>{testText}</p>
        </Card>
      );

      expect(container.firstChild).toHaveStyle(
        `box-shadow: 0px 0px 1px 0px rgba(61,66,81,0.3),${shadow.md.value}`
      );
    });

    it('has the large box shadow for elevation = 3', async () => {
      const { container } = await axeRender(
        <Card elevation={3}>
          <p>{testText}</p>
        </Card>
      );

      expect(container.firstChild).toHaveStyle(
        `box-shadow: 0px 0px 1px 0px rgba(61,66,81,0.3),${shadow.lg.value}`
      );
    });

    it('has the extra large box shadow for elevation = 4', async () => {
      const { container } = await axeRender(
        <Card elevation={4}>
          <p>{testText}</p>
        </Card>
      );

      expect(container.firstChild).toHaveStyle(
        `box-shadow: 0px 0px 1px 0px rgba(61,66,81,0.3),${shadow.xl.value}`
      );
    });
  });

  describe('snapshots', () => {
    Object.keys(elevationMap)
      .map(key => parseInt(key, 10))
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      .forEach((elevation: any) => {
        it(`renders a card with elevation ${elevation}`, async () => {
          const { container } = await axeRender(
            <Card elevation={elevation}>
              <p>example content</p>
            </Card>
          );

          expect(container.firstChild).toMatchSnapshot();
        });

        if (elevation === 0) return; // no hover snapshot for level 0

        it(`renders a card with hoverElevation ${elevation}`, async () => {
          const { container } = await axeRender(
            <Card hoverElevation={elevation}>
              <p>example content</p>
            </Card>
          );

          expect(container.firstChild).toMatchSnapshot();
        });
      });
  });
});
