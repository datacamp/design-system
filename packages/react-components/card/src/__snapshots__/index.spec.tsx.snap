// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> headstones renders the content of the headstone in a box above the card 1`] = `
.emotion-1 {
  background-color: rgb(255,255,255);
  border-radius: 4px;
  box-shadow: 0 0 0 1px rgba(5,25,45,0.1);
  -webkit-transition: all 0.6s cubic-bezier(0.075,0.82,0.165,1);
  transition: all 0.6s cubic-bezier(0.075,0.82,0.165,1);
  margin-top: 20px;
  position: relative;
}

.emotion-0 {
  height: 40px;
  left: 32px;
  position: absolute;
  top: -20px;
}

<div
  class="emotion-1"
>
  <div
    class="emotion-0"
  >
    this is in the headstone
  </div>
  <p>
    example content
  </p>
</div>
`;

exports[`<Card /> snapshots renders a card with elevation 0 1`] = `
.emotion-0 {
  background-color: rgb(255,255,255);
  border-radius: 4px;
  box-shadow: 0 0 0 1px rgba(5,25,45,0.1);
  -webkit-transition: all 0.6s cubic-bezier(0.075,0.82,0.165,1);
  transition: all 0.6s cubic-bezier(0.075,0.82,0.165,1);
}

<div
  class="emotion-0"
>
  <p>
    example content
  </p>
</div>
`;

exports[`<Card /> snapshots renders a card with elevation 1 1`] = `
.emotion-0 {
  background-color: rgb(255,255,255);
  border-radius: 4px;
  box-shadow: 0px 0px 1px 0px rgba(5,25,45,0.3),0 1px 4px -1px rgba(5,25,45,0.3);
  -webkit-transition: all 0.6s cubic-bezier(0.075,0.82,0.165,1);
  transition: all 0.6s cubic-bezier(0.075,0.82,0.165,1);
}

<div
  class="emotion-0"
>
  <p>
    example content
  </p>
</div>
`;

exports[`<Card /> snapshots renders a card with elevation 2 1`] = `
.emotion-0 {
  background-color: rgb(255,255,255);
  border-radius: 4px;
  box-shadow: 0px 0px 1px 0px rgba(5,25,45,0.3),0 2px 4px -1px rgba(5,25,45,0.3);
  -webkit-transition: all 0.6s cubic-bezier(0.075,0.82,0.165,1);
  transition: all 0.6s cubic-bezier(0.075,0.82,0.165,1);
}

<div
  class="emotion-0"
>
  <p>
    example content
  </p>
</div>
`;

exports[`<Card /> snapshots renders a card with elevation 3 1`] = `
.emotion-0 {
  background-color: rgb(255,255,255);
  border-radius: 4px;
  box-shadow: 0px 0px 1px 0px rgba(5,25,45,0.3),0 3px 5px -1px rgba(5,25,45,0.3);
  -webkit-transition: all 0.6s cubic-bezier(0.075,0.82,0.165,1);
  transition: all 0.6s cubic-bezier(0.075,0.82,0.165,1);
}

<div
  class="emotion-0"
>
  <p>
    example content
  </p>
</div>
`;

exports[`<Card /> snapshots renders a card with elevation 4 1`] = `
.emotion-0 {
  background-color: rgb(255,255,255);
  border-radius: 4px;
  box-shadow: 0px 0px 1px 0px rgba(5,25,45,0.3),0 8px 12px -4px rgba(5,25,45,0.3);
  -webkit-transition: all 0.6s cubic-bezier(0.075,0.82,0.165,1);
  transition: all 0.6s cubic-bezier(0.075,0.82,0.165,1);
}

<div
  class="emotion-0"
>
  <p>
    example content
  </p>
</div>
`;

exports[`<Card /> snapshots renders a card with hoverElevation 1 1`] = `
.emotion-0 {
  background-color: rgb(255,255,255);
  border-radius: 4px;
  box-shadow: 0 0 0 1px rgba(5,25,45,0.1);
  -webkit-transition: all 0.6s cubic-bezier(0.075,0.82,0.165,1);
  transition: all 0.6s cubic-bezier(0.075,0.82,0.165,1);
}

.emotion-0:hover {
  box-shadow: 0px 0px 1px 0px rgba(5,25,45,0.3),0 1px 4px -1px rgba(5,25,45,0.3);
  -webkit-transform: translate(0,-1px);
  -ms-transform: translate(0,-1px);
  transform: translate(0,-1px);
}

<div
  class="emotion-0"
>
  <p>
    example content
  </p>
</div>
`;

exports[`<Card /> snapshots renders a card with hoverElevation 2 1`] = `
.emotion-0 {
  background-color: rgb(255,255,255);
  border-radius: 4px;
  box-shadow: 0 0 0 1px rgba(5,25,45,0.1);
  -webkit-transition: all 0.6s cubic-bezier(0.075,0.82,0.165,1);
  transition: all 0.6s cubic-bezier(0.075,0.82,0.165,1);
}

.emotion-0:hover {
  box-shadow: 0px 0px 1px 0px rgba(5,25,45,0.3),0 2px 4px -1px rgba(5,25,45,0.3);
  -webkit-transform: translate(0,-1px);
  -ms-transform: translate(0,-1px);
  transform: translate(0,-1px);
}

<div
  class="emotion-0"
>
  <p>
    example content
  </p>
</div>
`;

exports[`<Card /> snapshots renders a card with hoverElevation 3 1`] = `
.emotion-0 {
  background-color: rgb(255,255,255);
  border-radius: 4px;
  box-shadow: 0 0 0 1px rgba(5,25,45,0.1);
  -webkit-transition: all 0.6s cubic-bezier(0.075,0.82,0.165,1);
  transition: all 0.6s cubic-bezier(0.075,0.82,0.165,1);
}

.emotion-0:hover {
  box-shadow: 0px 0px 1px 0px rgba(5,25,45,0.3),0 3px 5px -1px rgba(5,25,45,0.3);
  -webkit-transform: translate(0,-1px);
  -ms-transform: translate(0,-1px);
  transform: translate(0,-1px);
}

<div
  class="emotion-0"
>
  <p>
    example content
  </p>
</div>
`;

exports[`<Card /> snapshots renders a card with hoverElevation 4 1`] = `
.emotion-0 {
  background-color: rgb(255,255,255);
  border-radius: 4px;
  box-shadow: 0 0 0 1px rgba(5,25,45,0.1);
  -webkit-transition: all 0.6s cubic-bezier(0.075,0.82,0.165,1);
  transition: all 0.6s cubic-bezier(0.075,0.82,0.165,1);
}

.emotion-0:hover {
  box-shadow: 0px 0px 1px 0px rgba(5,25,45,0.3),0 8px 12px -4px rgba(5,25,45,0.3);
  -webkit-transform: translate(0,-1px);
  -ms-transform: translate(0,-1px);
  transform: translate(0,-1px);
}

<div
  class="emotion-0"
>
  <p>
    example content
  </p>
</div>
`;
